### PR TITLE
Initial commit -- Adding calibration loss specific to segmentation

### PR DESCRIPTION
**Pull request by Bala93**
### Description

Model calibration has helped in developing reliable deep learning models. In this pull request, I have added a new loss function NACL (https://arxiv.org/abs/2303.06268, https://arxiv.org/abs/2401.14487) which has shown promising results for both discriminative and calibration in segmentation. 

**Future Plans:** Currently, MONAI has some of the alternative loss functions (Label Smoothing, and Focal Loss), but it doesn't have the calibration specific loss functions (https://arxiv.org/abs/2111.15430, https://arxiv.org/abs/2209.09641). Besides, these methods are better evaluated with calibration metrics, Expected Calibration Error (https://lightning.ai/docs/torchmetrics/stable/classification/calibration_error.html).

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/losses/segcalib.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        self.svls_layer.weight.data = svls_kernel_2d
+        self.svls_layer.weight.requires_grad = False
+
+    def forward(self, x):
+        return self.svls_layer(x) / self.svls_kernel.sum()
+
+
+class NACLLoss(_Loss):
+    """
+    Murugesan, Balamurali, et al.
```
Comment (ericspod) [inline monai/losses/segcalib.py:65]: Please add here a brief description of what the loss function does, what applications it's for, and a bit more description of what the arguments do and mean.

File: monai/losses/segcalib.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    def forward(self, inputs: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
+        loss_ce = self.ce(inputs, targets)
+
+        utargets = self.get_constr_target(targets)
+
+        if self.distance_type == "l1":
+            loss_conf = torch.abs(utargets - inputs).mean()
+
+        if self.distance_type == "l2":
+            loss_conf = (torch.abs(utargets - inputs) ** 2).mean()
```
Comment (ericspod) [inline monai/losses/segcalib.py:169]:
```suggestion
        if self.distance_type == "l1":
            loss_conf = utargets.sub(inputs).abs_().mean()
        elif self.distance_type == "l2":
            loss_conf = utargets.sub(inputs).pow_(2).abs_().mean()
```
Maybe a little faster?

File: monai/losses/segcalib.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        loss = loss_ce + self.alpha * loss_conf
+
+        return loss  # , loss_ce, loss_conf
```
Comment (ericspod) [inline monai/losses/segcalib.py:173]:
```suggestion
        return loss 
```

**Review by ericspod — COMMENTED** [review]
Thanks for the contribution! I had a few comments, and the CICD errors are easy fixes I think. The test fail is from accuracy it seems which can be fixed by changing the error.

File: monai/losses/segcalib.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    xy_grid = torch.stack([x_grid, y_grid], dim=-1).float()
+    mean = (ksize - 1) / 2.0
+    variance = sigma**2.0
+    gaussian_kernel = (1.0 / (2.0 * math.pi * variance + 1e-16)) * torch.exp(
+        -torch.sum((xy_grid - mean) ** 2.0, dim=-1) / (2 * variance + 1e-16)
+    )
+    return gaussian_kernel / torch.sum(gaussian_kernel)
+
+
+class GaussianFilter(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (Bala93) [reply monai/losses/segcalib.py:62]: Hi @ericspod, I have integrated the support for both 2d and 3d now. I can migrate to GaussianFilter and MeanFilter in some time. Thanks.

File: monai/losses/segcalib.py
Comment (ericspod) [inline monai/losses/segcalib.py:1]: I would rename this file to `nacl_loss.py` to match the naming of other loss files in this directory.

File: tests/test_nacl_loss.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+                            [0.2622, 0.5002, 0.1905, 0.1644],
+                            [0.6354, 0.0047, 0.1649, 0.7112],
+                        ],
+                    ]
+                ]
+            ),
+            "targets": torch.tensor([[[1, 2, 0, 1], [0, 2, 1, 2], [0, 0, 2, 1], [1, 1, 1, 2]]]),
+        },
+        1.0375,
+    ],
```
Comment (ericspod) [inline tests/test_nacl_loss.py:115]: This test appears to test the same thing as the previous? Does this add anything?

File: tests/test_nacl_loss.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                        ],
+                        [
+                            [0.0759, 0.0238, 0.5570, 0.1691],
+                            [0.2703, 0.7722, 0.1611, 0.6431],
+                            [0.8051, 0.6596, 0.4121, 0.1125],
+                            [0.5283, 0.6746, 0.5528, 0.7913],
+                        ],
+                    ]
+                ]
+            ),
```
Comment (ericspod) [inline tests/test_nacl_loss.py:80]: This tensor is the same as that in the previous test, we can save some lines by putting this into a temporary variable and used it here and in the previous test.

File: monai/losses/segcalib.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ericspod) [reply monai/losses/segcalib.py:63]: If you can move to our existing definitions with the same results that would be great, thanks.

**Comment by ericspod** [comment]
Hi @Bala93 I have a few further comments, the fixes for the `mypy` issues should be minor and the DCO page tells you how to address that. Thanks!

File: tests/test_nacl_loss.py (hunk 5 above)
Comment (Bala93) [reply tests/test_nacl_loss.py:115]: Sure, got it. I will remove one of the case, it is just different values for the input and target tensors.

File: tests/test_nacl_loss.py (hunk 6 above)
Comment (Bala93) [reply tests/test_nacl_loss.py]: Fixed this one.

**Comment by Bala93** [comment]
> Hi @Bala93 I have a few further comments, the fixes for the `mypy` issues should be minor and the DCO page tells you how to address that. Thanks!

Hi @ericspod, Thanks for your feedback. I have addressed all the issues. I will work on using the monai functions (for the gaussian, and mean) in some time. Until then, is it possible to merge it with the main branch and make the loss function available ? Thank you.

File: docs/source/losses.rst [hunk 7]
```diff
@@ -93,6 +93,11 @@ Segmentation Losses
 .. autoclass:: SoftDiceclDiceLoss
     :members:
 
+`NACLLoss`
+~~~~~~~~~~~~~~~~~~~~
```
Comment (KumoLiu) [inline docs/source/losses.rst:97]:
```suggestion
`NACLLoss`
~~~~~~~~~~
```

File: monai/losses/nacl_loss.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+from monai.utils import pytorch_after
+
+
+def get_mean_kernel_2d(ksize: int = 3) -> torch.Tensor:
+    mean_kernel = torch.ones([ksize, ksize]) / (ksize**2)
+    return mean_kernel
+
+
+def get_mean_kernel_3d(ksize: int = 3) -> torch.Tensor:
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:30]: Perhaps we can combine these two to one function and move it under this [folder](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/convutils.py) to make it to be a more common used utils. What do you think?

cc @ericspod @Nic-Ma

File: monai/losses/nacl_loss.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    mean_kernel = torch.ones([ksize, ksize, ksize]) / (ksize**3)
+    return mean_kernel
+
+
+def get_gaussian_kernel_2d(ksize: int = 3, sigma: float = 1.0) -> torch.Tensor:
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:35]: Also for these two gaussian kernel functions.

File: monai/losses/nacl_loss.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+    mean = (ksize - 1) / 2.0
+    variance = sigma**2.0
+    gaussian_kernel: torch.Tensor = (1.0 / (2.0 * math.pi * variance + 1e-16)) * torch.exp(
+        -torch.sum((xyz_grid - mean) ** 2.0, dim=-1) / (2 * variance + 1e-16)
+    )
+    gaussian_kernel = gaussian_kernel / torch.sum(gaussian_kernel)
+    return gaussian_kernel
+
+
+class GaussianFilter(torch.nn.Module):
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:65]: As @ericspod suggested, can we integrate these class into the existing one?

File: monai/losses/nacl_loss.py [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+            )
+            self.svls_layer.weight.data = svls_kernel_3d
+            self.svls_layer.weight.requires_grad = False
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        svls_normalized: torch.Tensor = self.svls_layer(x) / self.svls_kernel.sum()
+        return svls_normalized
+
+
+class MeanFilter(torch.nn.Module):
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:121]: And for this class, would it be better to move it under monai/networks/layers/simplelayers.py?

File: monai/losses/nacl_loss.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+        self.svls_layer: Any
+
+        if kernel_ops == "mean":
+            self.svls_layer = MeanFilter(dim=dim, ksize=kernel_size, channels=classes)
+        if kernel_ops == "gaussian":
+            self.svls_layer = GaussianFilter(dim=dim, ksize=kernel_size, sigma=sigma, channels=classes)
+
+        self.old_pt_ver = not pytorch_after(1, 10)
+
+    # def ce(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:231]: please remove the comment lines here.

File: monai/losses/nacl_loss.py [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+    #             f"Multichannel targets are not supported in this older Pytorch version {torch.__version__}. "
+    #             "Using argmax (as a workaround) to convert target to a single channel."
+    #         )
+    #         target = torch.argmax(target, dim=1)
+    #     elif not torch.is_floating_point(target):
+    #         target = target.to(dtype=input.dtype)
+
+    #     return self.cross_entropy(input, target)  # type: ignore[no-any-return]
+
+    def get_constr_target(self, mask: torch.Tensor) -> torch.Tensor:
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:253]: Could you please add a small docstring here to demenstrate what's this function for?

File: tests/test_nacl_loss.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+                [0.8051, 0.6596, 0.4121, 0.1125],
+                [0.5283, 0.6746, 0.5528, 0.7913],
+            ],
+        ]
+    ]
+)
+targets = torch.tensor([[[1, 1, 1, 1], [1, 1, 1, 0], [0, 0, 0, 0], [0, 0, 0, 0]]])
+
+TEST_CASES = [
+    [{"classes": 3, "dim": 2}, {"inputs": inputs, "targets": targets}, 1.1820],
```
Comment (KumoLiu) [inline tests/test_nacl_loss.py:49]: Could you please include more tests for other supported kernel_ops and arguments?

**Review by KumoLiu — COMMENTED** [review]
Thanks for the contribution! Overall looks good to me. Several minor inline comments.

File: monai/losses/nacl_loss.py (hunk 8 above)
Comment (ericspod) [reply monai/losses/nacl_loss.py:30]:
```python
def get_mean_kernel(ksize: int = 3, dim: int = 2) -> torch.Tensor:
    return torch.ones([ksize] * dim) / (ksize**dim)
```
Something like this.

Comment (Bala93) [reply monai/losses/nacl_loss.py:30]: Sure, got the idea. Will work on this as well as for the gaussian (https://github.com/Project-MONAI/MONAI/pull/7819#discussion_r1705520061)

**Comment by Bala93** [comment]
Hi @KumoLiu @ericspod. I have addressed the comments with the latest commit. Thank you.

File: monai/losses/nacl_loss.py (hunk 10 above)
Comment (Bala93) [reply monai/losses/nacl_loss.py]: I have a doubt with GaussianFilter implementation. In the MONAI implementation, we have sigma and truncate. Is it possible for me to use the kernel size and sigma to decide the truncate value ? https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L554 This might not have much impact with the objective function, but would be good adding it.

Comment (KumoLiu) [reply monai/losses/nacl_loss.py]: The truncation range typically determines the kernel size; the kernel size must be large enough to accommodate the majority of the Gaussian distribution's energy. Of course, you can also extend the current function; just be careful to ensure compatibility. Thanks!

File: monai/losses/nacl_loss.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+            sigma: sigma of gaussian
+        """
+
+        super().__init__()
+
+        if kernel_ops not in ["mean", "gaussian"]:
+            raise ValueError("Kernel ops must be either mean or gaussian")
+
+        if dim not in [2, 3]:
+            raise ValueError("Supoorts 2d and 3d")
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:66]:
```suggestion
            raise ValueError(f"Support 2d and 3d, got dim={dim}.")
```

File: monai/losses/nacl_loss.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+        if distance_type not in ["l1", "l2"]:
+            raise ValueError("Distance type must be either L1 or L2")
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:69]:
```suggestion
            raise ValueError(f"Distance type must be either L1 or L2, got {distance_type}")
```

File: monai/losses/nacl_loss.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+            self.svls_layer = GaussianFilter(spatial_dims=dim, sigma=sigma)
+
+    def get_constr_target(self, mask: torch.Tensor) -> torch.Tensor:
+        """
+        Converts the mask to one hot represenation and applies the spatial filter.
+        """
+        rmask: torch.Tensor
+
+        if self.dim == 2:
+            oh_labels = F.one_hot(mask.to(torch.int64), num_classes=self.nc).contiguous().permute(0, 3, 1, 2).float()
```
Comment (KumoLiu) [inline monai/losses/nacl_loss.py:92]: Could you please add more description for the input shape for this function and also for the forward, such as the definition of dimension (we often assume all non-transform components assume input data shape [batch x channel x spatial_dims]) and also need to be one-hot or not. 

Here is an example in the `DiceCELoss`.
https://github.com/Project-MONAI/MONAI/blob/6c23fd06fc11667beedd0ba730d4104076a8db2d/monai/losses/dice.py#L776-L784

**Review by KumoLiu — COMMENTED** [review]
Thanks for the quick update, few further comments inline.

**Comment by ericspod** [comment]
Looks good now to me with the last minor comments addressed.

File: monai/losses/nacl_loss.py (hunk 17 above)
Comment (Bala93) [reply monai/losses/nacl_loss.py:92]: In my current implementation, I am treating it like typical Cross entropy loss in Pytorch. Ex: Predictions will be BNHW(D) and labels will be BHW(D). B - Batch size, N - No. of classes, HW(D) - Image dimensions. Can I stick with this and add this in docstring ?

Comment (KumoLiu) [reply monai/losses/nacl_loss.py:92]: Sure, it's just align with monai's assumption.

Comment (Bala93) [reply monai/losses/nacl_loss.py:98]: Have updated the docstring in both the functions following the example.

**Review by KumoLiu — APPROVED** [review]
Thanks for the quick update, looks good to me now. Will trigger more tests and try to merge this one.
